inotify/inotify_linux.go

```diff
@@ -165,37 +225,11 @@ func (w *Watcher) readEvents() {
 			continue
 		}
 
-		var offset uint32
-		// We don't know how many events we just read into the buffer
-		// While the offset points to at least one whole event...
-		for offset <= uint32(n-syscall.SizeofInotifyEvent) {
-			// Point "raw" to the event in the buffer
-			raw := (*syscall.InotifyEvent)(unsafe.Pointer(&buf[offset]))
-			event := new(Event)
-			event.Mask = uint32(raw.Mask)
-			event.Cookie = uint32(raw.Cookie)
-			nameLen := uint32(raw.Len)
-			// If the event happened to the watched directory or the watched file, the kernel
-			// doesn't append the filename to the event, but we would like to always fill the
-			// the "Name" field with a valid filename. We retrieve the path of the watch from
-			// the "paths" map.
-			w.mu.Lock()
-			name, ok := w.paths[int(raw.Wd)]
-			w.mu.Unlock()
-			if ok {
-				event.Name = name
-				if nameLen > 0 {
-					// Point "bytes" at the first byte of the filename
-					bytes := (*[syscall.PathMax]byte)(unsafe.Pointer(&buf[offset+syscall.SizeofInotifyEvent]))
-					// The filename is padded with NUL bytes. TrimRight() gets rid of those.
-					event.Name += "/" + strings.TrimRight(string(bytes[0:nameLen]), "\000")
-				}
-				// Send the event on the events channel
-				w.Event <- event
-			}
-			// Move to the next event in the buffer
-			offset += syscall.SizeofInotifyEvent + nameLen
-		}
+		cond.L.Lock()
+		bufs = append(bufs, bufn{buf, n})
```

Comment:
This change appears to be repeatedly reading into the same buffer (`buf`), then adding a reference to the same memory to a queue (`bufs`) which is asynchronously processed. Doesn't that mean the `syscall.Read(w.fd, buf[:])` call is overwriting data that is queued to be processed?

If I'm reading this correctly, this change is "improving" memory use in a way that breaks the data being processed.

Reply:
is it possible that the root cause might be a slow event reader that never drains the file event channel?

Reply:
yes, there are two channel are used: one's size is 16,another is 1.
in fact， i've tried to do two other changes：
the first is，i had make it  a goroutine
https://github.com/kubernetes/kubernetes/blob/3fc1bc76a95d8a6a6cd33394526f00c00a98ce07/vendor/github.com/google/cadvisor/container/raw/watcher.go#L230
another  is：i change the bufsize from SizeofInotifyEvent * 4096 to SizeofInotifyEvent * 40960, but that‘s also no use. 
the channels are always blocked finnally

Reply:
> i change the bufsize from SizeofInotifyEvent * 4096 to SizeofInotifyEvent * 40960, but that‘s also no use.

I don't think that is the problematic buffer... it seems like w.Event blocking is the issue

Reply:
right, i think so too for now